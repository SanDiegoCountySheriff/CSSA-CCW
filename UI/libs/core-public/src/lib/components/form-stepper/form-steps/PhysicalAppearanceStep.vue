<template>
  <div>
    <v-form
      ref="form"
      v-model="state.valid"
    >
      <v-subheader class="sub-header font-weight-bold">
        {{ $t(' Physical Appearance') }}
      </v-subheader>
      <v-row class="ml-5">
        <v-col
          cols="12"
          lg="6"
        >
          <v-text-field
            dense
            outlined
            :label="$t('Height feet')"
            type="number"
            :rules="[
              v =>
                (v > 0 && v < 10) ||
                $t('Height feet must be greater than 0 and less than 10'),
              v => !!v || $t('Height feet is required'),
            ]"
            v-model="completeApplication.physicalAppearance.heightFeet"
          >
            <template #prepend>
              <v-icon
                x-small
                color="error"
              >
                mdi-star
              </v-icon>
            </template>
          </v-text-field>
        </v-col>

        <v-col
          cols="12"
          lg="6"
        >
          <v-text-field
            dense
            outlined
            type="number"
            :label="$t('Height inches')"
            :rules="[
              v => !!v || $t('Height inches is required'),
              v => v < 12 || $t('Height inches must be less than 11'),
            ]"
            v-model="completeApplication.physicalAppearance.heightInch"
          >
            <template #prepend>
              <v-icon
                x-small
                color="error"
              >
                mdi-star
              </v-icon>
            </template>
          </v-text-field>
        </v-col>

        <v-col
          cols="12"
          lg="6"
        >
          <v-text-field
            dense
            outlined
            persistent-hint
            type="number"
            :hint="$t('Enter weight in pounds(lbs)')"
            :label="$t('Weight')"
            :rules="[
              v => !!v || $t('Weight is required'),
              v =>
                (v > 0 && v < 2000) ||
                $t('Weight must greater than 0 and less than 2000'),
            ]"
            v-model="completeApplication.physicalAppearance.weight"
          >
            <template #prepend>
              <v-icon
                x-small
                color="error"
              >
                mdi-star
              </v-icon>
            </template>
          </v-text-field>
        </v-col>

        <v-col
          cols="12"
          lg="6"
        >
          <v-select
            dense
            outlined
            v-model="completeApplication.physicalAppearance.hairColor"
            :value="completeApplication.physicalAppearance.hairColor"
            :items="hairColors"
            :label="$t('Hair Color')"
            :rules="[v => !!v || $t(' Hair color is required')]"
          >
            <template #prepend>
              <v-icon
                x-small
                color="error"
              >
                mdi-star
              </v-icon>
            </template>
          </v-select>
        </v-col>
        <v-col
          cols="12"
          lg="6"
        >
          <v-select
            dense
            outlined
            v-model="completeApplication.physicalAppearance.eyeColor"
            :value="completeApplication.physicalAppearance.eyeColor"
            :items="eyeColors"
            :label="$t('Eye Color')"
            :rules="[v => !!v || $t('Eye color is required')]"
          >
            <template #prepend>
              <v-icon
                x-small
                color="error"
              >
                mdi-star
              </v-icon>
            </template>
          </v-select>
        </v-col>
      </v-row>

      <v-subheader class="sub-header font-weight-bold">
        {{ $t(' Gender ') }}
      </v-subheader>
      <v-row class="ml-5">
        <v-col
          cols="12"
          lg="6"
        >
          <v-radio-group
            :label="'Gender'"
            v-model="completeApplication.physicalAppearance.gender"
            row
          >
            <v-radio
              :label="'Male'"
              :value="'male'"
            />
            <v-radio
              :label="'Female'"
              :value="'female'"
            />
          </v-radio-group>
          <v-alert
            dense
            outlined
            type="error"
            v-if="!completeApplication.physicalAppearance.gender"
          >
            {{ $t('Must select a gender') }}
          </v-alert>
        </v-col>
        <v-col
          cols="12"
          lg="6"
        >
          <v-textarea
            dense
            outlined
            counter
            maxlength="1000"
            v-model="completeApplication.physicalAppearance.physicalDesc"
            :label="$t('Physical Description')"
            :rules="[
              v =>
                !v ||
                (v && v.length <= 1000) ||
                $t('Maximum 1000 characters are allowed'),
            ]"
            clearable
          />
        </v-col>
      </v-row>
    </v-form>
    <v-divider />
    <FormButtonContainer
      :valid="state.valid"
      @submit="handleSubmit"
      @save="saveMutation.mutate"
      @back="handlePreviousSection"
      @cancel="router.push('/')"
    />
    <v-snackbar
      :value="state.snackbar"
      :timeout="3000"
      bottom
      color="error"
      outlined
    >
      {{ $t('Section update unsuccessful please try again.') }}
    </v-snackbar>
    <v-snackbar
      :value="state.formError"
      :timeout="3000"
      bottom
      color="error"
      outlined
    >
      {{ $t('You must select a gender') }}
    </v-snackbar>
  </div>
</template>

<script setup lang="ts">
import FormButtonContainer from '@shared-ui/components/containers/FormButtonContainer.vue';
import { reactive } from 'vue';
import { useCompleteApplicationStore } from '@shared-ui/stores/completeApplication';
import { useMutation } from '@tanstack/vue-query';
import { useRouter } from 'vue-router/composables';
import { eyeColors, hairColors } from '@shared-utils/lists/defaultConstants';

interface FormStepFourProps {
  handleNextSection: () => void;
  handlePreviousSection: CallableFunction;
}

const props = withDefaults(defineProps<FormStepFourProps>(), {
  handleNextSection: () => null,
});

const state = reactive({
  valid: false,
  snackbar: false,
  formError: false,
});
const completeApplicationStore = useCompleteApplicationStore();
const completeApplication =
  completeApplicationStore.completeApplication.application;
const router = useRouter();

const updateMutation = useMutation({
  mutationFn: () => {
    return completeApplicationStore.updateApplication();
  },
  onSuccess: () => {
    completeApplication.currentStep = 5;
    props.handleNextSection();
  },
  onError: () => {
    state.snackbar = true;
  },
});

const saveMutation = useMutation({
  mutationFn: () => {
    return completeApplicationStore.updateApplication();
  },
  onSuccess: () => {
    router.push('/');
  },
  onError: () => {
    state.snackbar = true;
  },
});

function handleSubmit() {
  if (!completeApplication.physicalAppearance.gender) {
    state.formError = true;
  } else {
    updateMutation.mutate();
  }
}
</script>
