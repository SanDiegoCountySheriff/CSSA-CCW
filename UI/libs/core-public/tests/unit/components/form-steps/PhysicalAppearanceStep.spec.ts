// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
import PhysicalAppearanceStep from '@core-public/components/form-stepper/form-steps/PhysicalAppearanceStep.vue';
import Vuetify from 'vuetify';
import { createTestingPinia } from '@pinia/testing';
import { createLocalVue, mount } from '@vue/test-utils';

const pinia = createTestingPinia();
const localVue = createLocalVue();
const tMock = {
  $t: text => text,
};

describe('FormStepFour', () => {
  let vuetify;
  let wrapper;

  beforeEach(() => {
    vuetify = new Vuetify();
    wrapper = mount(PhysicalAppearanceStep, {
      localVue,
      vuetify,
      pinia,
      mocks: tMock,
      propsData: {
        handleNextSection: () => null,
      },
    });
  });

  it('Should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render all the inputs', () => {
    expect(wrapper.findAllComponents('.v-text-field')).toHaveLength(6);
  });
});
