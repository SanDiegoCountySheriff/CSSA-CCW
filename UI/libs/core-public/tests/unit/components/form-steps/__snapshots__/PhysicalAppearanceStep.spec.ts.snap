// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormStepFour Should match the snapshot 1`] = `
<div>
  <form novalidate="novalidate" class="v-form">
    <div class="v-subheader sub-header font-weight-bold theme--light">
      Physical Appearance
    </div>
    <div class="row ml-5">
      <div class="col-lg-6 col-12">
        <div class="v-input v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__prepend-outer"><i aria-hidden="true" class="v-icon notranslate mdi mdi-star theme--light error--text" style="font-size: 12px;"></i></div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span class="notranslate">​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="input-4" class="v-label theme--light" style="left: 0px; position: absolute;">Height feet</label><input id="input-4" type="number"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-lg-6 col-12">
        <div class="v-input v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__prepend-outer"><i aria-hidden="true" class="v-icon notranslate mdi mdi-star theme--light error--text" style="font-size: 12px;"></i></div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span class="notranslate">​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="input-8" class="v-label theme--light" style="left: 0px; position: absolute;">Height inches</label><input id="input-8" type="number"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-lg-6 col-12">
        <div class="v-input v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__prepend-outer"><i aria-hidden="true" class="v-icon notranslate mdi mdi-star theme--light error--text" style="font-size: 12px;"></i></div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span class="notranslate">​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="input-12" class="v-label theme--light" style="left: 0px; position: absolute;">Weight</label><input id="input-12" type="number"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                  <div class="v-messages__message">Enter weight in pounds(lbs)</div>
                </transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-lg-6 col-12">
        <div class="v-input v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
          <div class="v-input__prepend-outer"><i aria-hidden="true" class="v-icon notranslate mdi mdi-star theme--light error--text" style="font-size: 12px;"></i></div>
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-16" class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span class="notranslate">​</span></legend>
              </fieldset>
              <div class="v-select__slot"><label for="input-16" class="v-label theme--light" style="left: 0px; position: absolute;">Hair Color</label>
                <div class="v-select__selections"><input id="input-16" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-lg-6 col-12">
        <div class="v-input v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
          <div class="v-input__prepend-outer"><i aria-hidden="true" class="v-icon notranslate mdi mdi-star theme--light error--text" style="font-size: 12px;"></i></div>
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-22" class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span class="notranslate">​</span></legend>
              </fieldset>
              <div class="v-select__slot"><label for="input-22" class="v-label theme--light" style="left: 0px; position: absolute;">Eye Color</label>
                <div class="v-select__selections"><input id="input-22" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="v-subheader sub-header font-weight-bold theme--light">
      Gender
    </div>
    <div class="row ml-5">
      <div class="col-lg-6 col-12">
        <div class="v-input theme--light v-input--selection-controls v-input--radio-group v-input--radio-group--row">
          <div class="v-input__control">
            <div class="v-input__slot" style="height: auto;">
              <div role="radiogroup" aria-labelledby="input-29" class="v-input--radio-group__input">
                <legend id="input-29" class="v-label theme--light" style="left: 0px; position: relative;">Gender</legend>
                <div class="v-radio theme--light">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate mdi mdi-radiobox-blank theme--light"></i><input aria-checked="false" id="input-30" role="radio" type="radio" name="radio-29" value="male">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div><label for="input-30" class="v-label theme--light" style="left: 0px; position: relative;">Male</label>
                </div>
                <div class="v-radio theme--light">
                  <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate mdi mdi-radiobox-blank theme--light"></i><input aria-checked="false" id="input-32" role="radio" type="radio" name="radio-29" value="female">
                    <div class="v-input--selection-controls__ripple"></div>
                  </div><label for="input-32" class="v-label theme--light" style="left: 0px; position: relative;">Female</label>
                </div>
              </div>
            </div>
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
        <div role="alert" class="v-alert v-sheet v-sheet--outlined theme--light v-alert--dense v-alert--outlined error--text">
          <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--light error--text"></i>
            <div class="v-alert__content">
              Must select a gender
            </div>
          </div>
        </div>
      </div>
      <div class="col-lg-6 col-12">
        <div class="v-input v-textarea v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span class="notranslate">​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="input-38" class="v-label theme--light" style="left: 0px; position: absolute;">Physical Description</label><textarea maxlength="1000" id="input-38" rows="5"></textarea></div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
              <div class="v-counter theme--light">0 / 1000</div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
  <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  <div class="form-btn-container">
    <div class="form-btn-inner">
      <div><button type="button" disabled="disabled" class="mt-3 v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"><span class="v-btn__content">
        Continue
      </span></button> <button type="button" disabled="disabled" class="mt-3 v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"><span class="v-btn__content">
        Save and Exit
      </span></button></div>
      <div><button type="button" class="mt-3 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--small warning mx-2"><span class="v-btn__content">
        Go back
      </span></button> <button type="button" class="mt-3 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--small error mr-2"><span class="v-btn__content">
        Cancel
      </span></button></div>
    </div>
  </div>
  <div class="v-snack v-snack--bottom">
    <transition-stub name="v-snack-transition">
      <div class="v-snack__wrapper v-sheet v-sheet--outlined theme--light error--text" style="display: none;">
        <div role="status" aria-live="polite" class="v-snack__content">
          Section update unsuccessful please try again.
        </div>
        <div class="v-snack__action "></div>
      </div>
    </transition-stub>
  </div>
  <div class="v-snack v-snack--bottom">
    <transition-stub name="v-snack-transition">
      <div class="v-snack__wrapper v-sheet v-sheet--outlined theme--light error--text" style="display: none;">
        <div role="status" aria-live="polite" class="v-snack__content">
          You must select a gender
        </div>
        <div class="v-snack__action "></div>
      </div>
    </transition-stub>
  </div>
</div>
`;
